File: prism-kotlin-sdk/identity/src/commonMain/kotlin/io.iohk.atala.prism.kotlin.identity/DIDUrl.kt
package io.iohk.atala.prism.kotlin.identity

import com.github.h0tk3y.betterParse.grammar.tryParseToEnd
import com.github.h0tk3y.betterParse.parser.ErrorResult
import com.github.h0tk3y.betterParse.parser.Parsed

data class DIDUrl(
    val did: DID,
    val path: List<String>,
    val parameters: Map<String, List<String>>,
    val fragment: String?
) {
    companion object {
        fun fromString(rawDidUrl: String): DIDUrl =
            when (val result = DIDParser.didUrlGrammar.tryParseToEnd(rawDidUrl)) {
                is Parsed -> result.value
                is ErrorResult -> {
                    println(result)
                    throw IllegalArgumentException("Invalid DID URL: $rawDidUrl")
                }
            }
    }

    val keyId: String?
        get() =
            if (path.size == 2 && path[0] == "keyId") {
                path[1]
            } else {
                null
            }
}

File: prism-sdk/identity/shared/src/main/scala/io/iohk/atala/prism/identity/DIDUrl.scala
package io.iohk.atala.prism.identity

import io.lemonlabs.uri.Url

// A DID URL identifies a specific resource, usually a part of DID document.
// See https://www.w3.org/TR/did-core/#did-url-syntax for more information.
final case class DIDUrl(
    did: DID,
    path: Vector[String],
    parameters: Map[String, Vector[String]],
    fragment: Option[String]
) {
  def keyIdOption: Option[String] = {
    path match {
      case Vector("keyId", keyId) => Some(keyId)
      case _ => None
    }
  }
}

object DIDUrl {
  sealed trait DIDUrlError
  final case class InvalidUrl(url: String) extends DIDUrlError
  final case object EmptyDidScheme extends DIDUrlError
  final case object EmptyDidSuffix extends DIDUrlError
  final case class InvalidDid(did: String) extends DIDUrlError

  def fromString(rawDidUrl: String): Either[DIDUrlError, DIDUrl] = {
    for {
      url <- Url.parseOption(rawDidUrl).toRight(InvalidUrl(rawDidUrl))
      scheme <- url.schemeOption.toRight(EmptyDidScheme)
      pathParts = url.path.parts
      didSuffix <- pathParts.headOption.toRight(EmptyDidSuffix)
      // Non-failing way to get potentially empty list's tail
      didUrlPath = pathParts.drop(1)
      rawDid = scheme + ":" + didSuffix
      did <- DID.fromString(rawDid).toRight(InvalidDid(rawDid))
    } yield DIDUrl(did, didUrlPath, url.query.paramMap, url.fragment)
  }
}
